### ISSUE TITLE
I'm getting ram error when join table

### ISSUE THREAD

**Issue by sefabalica** [ClickHouse/ClickHouse#59546]
Hello,
When joining 2 not very large tables (one is 1 GB and the other is 800 MB), even though I specify max_memory_usage=500000000000, I still get an error saying the RAM consumption is too high(475.02 GiB). The table creation, row and disk sizes, and join scripts are below. Is there a point I missed or is there a problem? I'll be happy if you can help me..

create table y_test(KimlikNo String,IlkKayitTarihi DateTime64(3),IlAdi LowCardinality(String)) ENGINE=Memory as
SELECT KimlikNo,IlkKayitTarihi,IlAdi FROM yolcu_all
WHERE  (IlkKayitTarihi > (now() + toIntervalMonth(-3))) AND (Kaynak = 'YOLCU')

create table k_test(KimlikNo String,GelisTarihi Nullable(DateTime64(3)),AyrilisTarihi Nullable(DateTime64(3)),TesisIl LowCardinality(String)) ENGINE=Memory as
SELECT KimlikNo,GelisTarihi,AyrilisTarihi,TesisIl FROM kon_all
WHERE   GelisTarihi > (now() + toIntervalMonth(-3))

select *,formatReadableSize(total_bytes) as boyut from system.tables where name in('k_test','y_test') 
y_test=>>**828.42 MiB 23.375.072 rows**
k_test=>>**1.10 GiB 19.365.196 rows**

select  count(distinct y.KimlikNo) FROM y_test y inner join k_test k on y.KimlikNo = k.KimlikNo AND y.IlAdi = k.TesisIl
WHERE (k.AyrilisTarihi < (y.IlkKayitTarihi + toIntervalDay(6)))
SETTINGS join_algorithm = 'parallel_hash',max_memory_usage=500000000000

**Received exception from server (version 22.11.4):
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (for query) exceeded: would use 475.02 GiB (attempt to allocate chunk of 34359738368 bytes), maximum: 465.66 GiB. OvercommitTracker decision: Memory overcommit isn't used. OvercommitTracker isn't set.: While executing JoiningTransform. (MEMORY_LIMIT_EXCEEDED)**

**Comment by Algunenano** [comment]
22.11 is out of support. Please upgrade

**Comment by sefabalica** [comment]
Hello,
Except for the Clickhouse cluster, I pulled the same data to a test server (64GB Ram) with Ch 23.10 installed. I changed the engine and the size of the tables decreased. I can now choose the join algorithm in grace_hash. I get an error when I try it with both grace_hash and parallel_hash. When you change the Join type to INNER ANY JOIN, it gives an answer, although it is not very short (it gives incomplete data because there is no cartesian multiplication). However, when I select INNER JOIN, I get an error. What should I do?

create table y_test(KimlikNo String,IlkKayitTarihi DateTime64(3),IlAdi LowCardinality(String)) ENGINE=**MergeTree** Order By KimlikNo as 
SELECT KimlikNo,IlkKayitTarihi,IlAdi FROM remote('X.X.X.X', default.yolcu_all, 'default', 'XXX') WHERE  (IlkKayitTarihi > (now() + toIntervalMonth(-3))) AND (Kaynak = 'YOLCU ŞAHIS')

create table k_test(KimlikNo String,GelisTarihi Nullable(DateTime64(3)),AyrilisTarihi Nullable(DateTime64(3)),TesisIl LowCardinality(String)) ENGINE=**MergeTree** Order By KimlikNo as SELECT KimlikNo,GelisTarihi,AyrilisTarihi,TesisIl FROM remote('X.X.X.X', default.kon_all, 'default', 'XXX') WHERE   GelisTarihi > (now() + toIntervalMonth(-3))

select total_rows,formatReadableSize(total_bytes) as boyut from system.tables where name in('k_test','y_test') 
Row 1:
total_rows: 19199648
boyut:      **280.35 MiB**
Row 2:
total_rows: 22367708
boyut:      **208.50 MiB**

select  count(distinct y.KimlikNo) FROM y_test y inner join k_test k on y.KimlikNo = k.KimlikNo AND y.IlAdi = k.TesisIl
WHERE (k.AyrilisTarihi < (y.IlkKayitTarihi + toIntervalDay(6)))
SETTINGS join_algorithm = '**parallel_hash**' 

Received exception from server (version 23.10.5):
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (total) exceeded: would use 57.31 GiB (attempt to allocate chunk of 8589934592 bytes), maximum: 56.47 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.: While executing JoiningTransform. (MEMORY_LIMIT_EXCEEDED)


select  count(distinct y.KimlikNo) FROM y_test y inner join k_test k on y.KimlikNo = k.KimlikNo AND y.IlAdi = k.TesisIl
WHERE (k.AyrilisTarihi < (y.IlkKayitTarihi + toIntervalDay(6)))
SETTINGS join_algorithm = '**grace_hash**' 

Received exception from server (version 23.10.5):
Code: 241. DB::Exception: Received from localhost:9000. DB::Exception: Memory limit (total) exceeded: would use 61.98 GiB (attempt to allocate chunk of 1073741824 bytes), maximum: 56.47 GiB. OvercommitTracker decision: Query was selected to stop by OvercommitTracker.: While executing JoiningTransform. (MEMORY_LIMIT_EXCEEDED)

select  count(distinct y.KimlikNo) FROM y_test y **inner any join** k_test k on y.KimlikNo = k.KimlikNo AND y.IlAdi = k.TesisIl
WHERE (k.AyrilisTarihi < (y.IlkKayitTarihi + toIntervalDay(6)))
SETTINGS join_algorithm = 'grace_hash'

 uniqExact(KimlikNo) 
           154413 

1 row in set. Elapsed: 8.434 sec. Processed 41.57 million rows, 1.08 GB (4.93 million rows/s., 128.58 MB/s.)
Peak memory usage: 1.65 GiB.
